packages/frontend/src/components/json_import.tsx

```diff
@@ -0,0 +1,106 @@
+import { createSignal } from "solid-js";
+import type { Document } from "../api";
+import "./json_import.css";
+
+interface JsonImportProps<T extends string> {
+    onImport: (data: Document<T>) => void;
+    validate?: (data: Document<T>) => boolean | string;
+}
+
+export const JsonImport = <T extends string>({ onImport, validate }: JsonImportProps<T>) => {
+    const [error, setError] = createSignal<string | null>(null);
+    const [importValue, setImportValue] = createSignal("");
+
+    const handleError = (e: unknown) => {
+        setError(e instanceof Error ? e.message : "Unknown error occurred");
+    };
+
+    const validateAndImport = (jsonString: string) => {
+        try {
+            const data = JSON.parse(jsonString);
+
+            // Run custom validation if provided
+            if (validate) {
+                const validationResult = validate(data);
+                if (typeof validationResult === "string") {
+                    setError(validationResult);
+                    return;
+                }
+            }
+
+            // Clear any previous errors and import
+            setError(null);
+            onImport(data);
+            setImportValue(""); // Clear paste area after successful import
+        } catch (e) {
+            handleError(e);
+        }
+    };
+
+    // Handle file upload
+    const handleFileUpload = async (event: Event) => {
+        const input = event.target as HTMLInputElement;
```

Comment:
You could save yourself the explicit type cast of `event.target` by typing `handleFileUpload` as follows: https://docs.solidjs.com/configuration/typescript#basics